Feat/slack phase ii

sandeshit opened the pull request:
## Changes

* Slack messages for:
1. Project creation
2. Project progress
3. Project status change

## This PR doesn't introduce any:

- [X] temporary files, auto-generated files or secret keys
- [X] n+1 queries
- [X] flake8 issues
- [X] `print`
- [X] typos
- [X] unwanted comments

## This PR contains valid:

- [ ] tests
- [ ] permission checks (tests here too)
- [ ] translations

sandeshit committed:
feat(slack): setup slack-sdk using webhooks

sandeshit committed:
feat(slack): add function for slack message

sandeshit committed:
feat(slack): base command for testing slack message

sandeshit committed:
feat(slack): convert all async functions to sync and use celery

sandeshit committed:
chore(slack): base setup for slack webclient

sandeshit committed:
chore(model): add slack_user_id and slack_thread_ts fields.

sandeshit committed:
feat(serializer): slack integration for all cases

sandeshit committed:
chore(test): test commands for all cases

sandeshit committed:
feat(test): add dummy slack client in mutation test

sandeshit committed:
feat(slack): boradcast thread message to the channel

sandeshit committed:
feat(slack): dynamic parent message

sandeshit committed:
feat(slack): add progress bar

sandeshit committed:
chore(slack): accept None type for tutorial_id and cover_image

tnagorra commented on .github/workflows/ci.yml:
```diff
@@ -26,6 +26,8 @@ env:
   COMPOSE_FILE: docker-compose.yaml:.github/gh-docker-compose.yaml
   SECRET_KEY: "ci-test-insecure-django-secret-key"
   COMPOSE_BAKE: true
+  SLACK_BOT_ENABLED: true
```
Why do we enable slack bot in CI?

tnagorra commented on apps/project/management/commands/test_slack_project_creation.py:
```diff
@@ -0,0 +1,25 @@
+    @typing.override
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--project_id",
```
```suggestion
            "--project-id",
```

tnagorra commented on apps/project/management/commands/test_slack_project_creation.py:
```diff
@@ -0,0 +1,25 @@
+    def handle(self, *args, **options):
+        project_id = options["project_id"]
+
+        project_creation_success_message.delay(project_id=project_id)
```
We don't need this to be executed asynchronously

tnagorra commented on apps/project/management/commands/test_slack_project_progress.py:
```diff
@@ -0,0 +1,39 @@
+    @typing.override
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--project_id",
```
```suggestion
            "--project-id",
```

tnagorra replied:
Or even `--project`

tnagorra commented on apps/project/management/commands/test_slack_project_progress.py:
```diff
@@ -0,0 +1,39 @@
+        try:
+            send_message_for_progress.delay(project_id=project_id)
+        except Exception as err:
+            raise CommandError("Failed to send slack message for progress") from err
```
Doesn't this run on async context? How can we catch exceptions here?

tnagorra commented on apps/project/management/commands/test_slack_project_progress.py:
```diff
@@ -0,0 +1,39 @@
+        slack_thread = project.slack_thread_ts
+
+        if not slack_thread:
+            raise CommandError(f"Project id={project_id} has no Slack thread. Message not sent.")
```
@thenav56 How should we handle case where there's no slack thread?

tnagorra commented on apps/project/models.py:
```diff
@@ -385,6 +385,9 @@ class Project(UserResource, FirebasePushResource):
     result_count = models.IntegerField[int, int](default=0)  # NOTE: All project have 0 in production database
 
+    # SLACK THREAD TS
+    slack_thread_ts = models.CharField(max_length=20, null=True, blank=True)
```
What is `thread_ts` ?

tnagorra replied:
Please add help text. Why was this named thread_ts?

tnagorra commented on apps/user/models.py:
```diff
@@ -43,6 +43,7 @@ class User(AbstractUser):
     )
 
     objects: CustomUserManager = CustomUserManager()  # type: ignore[reportAssignmentType]
+    slack_user_id = models.CharField(max_length=20, null=True, blank=False)
```
How are we planning to set this?
@thenav56 Does project manager handle these scenarios?

tnagorra commented on pyproject.toml:
```diff
@@ -45,6 +45,8 @@ dependencies = [
     "python-magic>=0.4.27",
     "pandas~=2.3.1",
+    "slack-sdk[aiohttp]>=3.36.0",
+    "aiohttp>=3.12.15",
```
Do we need asynchronous libraries for slack-sdk

cc: @thenav56

tnagorra commented on utils/common.py:
```diff
@@ -211,3 +215,10 @@ def to_groups[T](features: list[T], group_size: int, start_index: int = 100):
+    url = image_file.url
+    if isinstance(default_storage, FileSystemStorage):
+        return f"{urlunparse(settings.APP_DOMAIN)}{url}"
+    return url
```
Please use get_absolute_uri instead. We don't need this function.

tnagorra commented on main/settings.py:
```diff
@@ -510,6 +514,13 @@ def urlparse(value) -> ParseResult:
     SENTRY_CONFIG.init_sentry()
 
+# Slack
+SLACK_BOT_ENABLED = env("SLACK_BOT_ENABLED")
```
When SLACK_BOT_ENABLE=false, we should still log the message to the console.

tnagorra commented on apps/project/management/commands/test_slack_project_creation.py:
We should merge all the commands into one command and use sub-command for each action.

tnagorra commented on apps/project/management/commands/test_slack_status_update.py:
```diff
@@ -0,0 +1,40 @@
+        except Exception as err:
+            raise CommandError("Failed to send slack message for status update") from err
+
+        self.stdout.write(self.style.SUCCESS("Slack message sent successfully"))
```
Same as above.

tnagorra commented on apps/project/serializers.py:
```diff
@@ -502,4 +508,14 @@ def update(self, instance: Project, validated_data: dict[str, typing.Any]) -> Pr
+            project_creation_success_message.delay(project_id=instance.pk)
+
+        if instance.created_by != requesting_user:
+            project_status_update_message.delay(project_id=instance.pk)
```
We should always trigger slack message action on status change.
If there is no base message, we can send the message to that thread.
Else, we need to create a base message first (then send the message to the thread)

> A base message defines the project details.
> We also need to update the base message with the updated/latest metadata like
> project name, progress, status, etc.

We also a "by XXX" message for who initiated the action.

tnagorra commented on apps/project/tasks.py:
```diff
@@ -1,10 +1,41 @@
+    project_type: int
+    tutorial_id: int | None
+    requesting_organization: str
+    created_by: User
```
Didn't we have an error while we were passing Project object.
Does that same error occur with User?

tnagorra commented on apps/project/tasks.py:
```diff
@@ -62,3 +93,418 @@ def generate_project_exports(
+    ) -> dict:
+        def generate_progress_bar(progress: int, bar_length: int = 6) -> str:
+            filled_length = int(round(bar_length * progress // 100))
+            bar = "⬛" * filled_length + "⬜" * (bar_length - filled_length)
```
Can you use green and white block for progress.

tnagorra commented on apps/project/tasks.py:
```diff
@@ -62,3 +93,418 @@ def generate_project_exports(
+        is_private: bool = True,
+    ) -> dict:
+        def generate_progress_bar(progress: int, bar_length: int = 6) -> str:
+            filled_length = int(round(bar_length * progress // 100))
```
Please use floor instead of round.
Any reason to use rounder after doing a floor division?

tnagorra commented on apps/project/tasks.py:
```diff
@@ -62,3 +93,418 @@ def generate_project_exports(
+        progress: int,
+        is_private: bool = True,
+    ) -> dict:
+        def generate_progress_bar(progress: int, bar_length: int = 6) -> str:
```
Let's use bar length of 10

tnagorra commented on apps/project/tasks.py:
```diff
@@ -62,3 +93,418 @@ def generate_project_exports(
+
+        if tutorial_id is not None:
+            tutorial_text = (
+                f"Tutorial: <{manager_dashboard_url}/tutorials/{tutorial_id}/edit|Tutorial Link>\n"
```
Let's have a utility function or class to generate links for manager dashboard.
Please check example for firebase urls.

tnagorra commented on apps/project/tasks.py:
```diff
@@ -62,3 +93,418 @@ def generate_project_exports(
+                f"Tutorial ID: {tutorial_id}\n"
+            )
+        else:
+            tutorial_text = "Tutorial: _No tutorial linked_\n"
```
```suggestion
            tutorial_text = "Tutorial ID: _N/A_\n"
```

sandeshit committed:
refactor(slack): implement final design changes

tnagorra commented on apps/project/serializers.py:
```diff
@@ -597,4 +602,14 @@ def update(self, instance: Project, validated_data: dict[str, typing.Any]) -> Pr
+
+        if (old_status_enum == Project.Status.PUBLISHED) and old_status_enum != updated_project.status_enum:
+            status_update_message.delay(project_id=instance.pk)
+
```
Let's simplify this logic. We can even combine these together.

tnagorra commented on apps/project/management/commands/test_slack_message.py:
```diff
@@ -0,0 +1,43 @@
+
+
+class Command(BaseCommand):
+    help = "Test sending a Slack progress message for a given project"
```
```suggestion
    help = "Send slack message for a given project"
```

tnagorra commented on apps/project/management/commands/test_slack_message.py:
```diff
@@ -0,0 +1,43 @@
+            "--project-id",
+            type=str,
+            required=True,
+            help="ID of the project to test",
```
```suggestion
            help="ID of the project",
```

tnagorra commented on apps/project/management/commands/test_slack_message.py:
```diff
@@ -0,0 +1,43 @@
+        subparsers = parser.add_subparsers(dest="event", help="Event to trigger", required=True)
+        subparsers.add_parser("project-creation", help="Send message for project creation")
+        subparsers.add_parser("status-update", help="Project status changed")
+        subparsers.add_parser("progress-change", help="Project progress change")
```
```suggestion
        subparsers.add_parser("project-creation", help="Send message for project creation")
        subparsers.add_parser("status-update", help="Send message for project status changed")
        subparsers.add_parser("progress-change", help="Send message for progress change")
```

tnagorra commented on apps/project/tasks.py:
```diff
@@ -74,3 +78,333 @@ def regenerate_global_project_assets():
+    )
+    client.update_slack_message(ts=ts, **update_message)
+
+
```
Let's move this logic to separate file.

tnagorra commented on apps/project/tasks.py:
```diff
@@ -74,3 +78,333 @@ def regenerate_global_project_assets():
+                Project.Status.PAUSED: ":hand:",
+                Project.Status.ARCHIVED: ":archived:",
+            }
+            return f"{label} {status_to_icon.get(status_enum, '')}".strip()
```
We need to make this typesafe

tnagorra commented on apps/project/tasks.py:
```diff
@@ -74,3 +78,333 @@ def regenerate_global_project_assets():
+                Project.Type.STREET: ":street:",
+                Project.Type.COMPLETENESS: ":completeness:",
+                Project.Type.VALIDATE_IMAGE: ":validate_image:",
+            }
```
We need to make this typesafe

tnagorra commented on apps/project/tasks.py:
```diff
@@ -74,3 +78,333 @@ def regenerate_global_project_assets():
+        project_url = Config.ManagerDashboardUrls.project_url(project_id=project.id)
+        website_url = Config.WebsiteKeys.project(firebase_id=project.firebase_id)
+
+        if progress == 100:
```
```suggestion
        if progress >= 100:
```

Just to be extra safe.

tnagorra commented on apps/project/tasks.py:
```diff
@@ -74,3 +78,333 @@ def regenerate_global_project_assets():
+        cls,
+        project: Project,
+    ) -> MapswipeSlack.MapswipeSlackMessageArgumentType:
+        header_text = f"{project.generate_name()} :lock:" if project.is_private else f"{project.generate_name()}"
```
```suggestion
        header_text = f"{project.generate_name()} :lock:" if project.is_private else project.generate_name()
```

sandeshit committed:
refactor(serializer): combine logic and cleanup

sandeshit committed:
refactor(tasks): separate tasks and messages

sandeshit committed:
feat(test): add dummy slack client in new tests

tnagorra committed:
fix(slack): remove calls to slack tasks and update migrations

- merge all the slack tasks into one
    - remove calling slack tasks
    - we need to move the triggers to the project background tasks
- remove "project-creation" action from slack test commands
    - status change should be enough
- add an environment to enable slack messages fallback to console
    - log to console if the environment variable is enabled
- remove slack mock
    - currently the slack mock does not work
- fix migration after the force squash
- refactor SlackMessage and break down into format helpers
